### PR TITLE
puts: Return EOF on error

### PR DESCRIPTION
**Pull request by agkaminski**
DONE: RTOS-503

## Description

Fixes https://github.com/phoenix-rtos/phoenix-rtos-project/issues/260

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

**Review by damianloew — DISMISSED** [review]
LGTM

File: stdio/file.c [hunk 1]
```diff
@@ -885,8 +885,11 @@ int puts(const char *s)
 
 int fputs_unlocked(const char *s, FILE *stream)
 {
-	int len = strlen(s);
-	return fwrite_unlocked(s, 1, len, stream);
+	size_t len = strlen(s);
+	if (fwrite_unlocked(s, 1, len, stream) != len) {
+		return EOF;
+	}
+	return (int)len;
```
Comment (badochov) [inline stdio/file.c:892]: `len > MAX_INT ? MAX_INT : len`

File: stdio/file.c [hunk 2]
```diff
@@ -885,8 +885,11 @@ int puts(const char *s)
 
 int fputs_unlocked(const char *s, FILE *stream)
 {
-	int len = strlen(s);
-	return fwrite_unlocked(s, 1, len, stream);
+	size_t len = strlen(s);
+	if (fwrite_unlocked(s, 1, len, stream) != len) {
```
Comment (badochov) [inline stdio/file.c:889]: I think the write should be retried until either len bytes is written or error is encoutered

**Review by badochov — CHANGES_REQUESTED** [review]

**Review by badochov — APPROVED** [review]
LGTM, maybe `puts_unlocked` should use this function to simplify the code a little bit

**Comment by agkaminski** [comment]
I don't think my change is causing `host-generic-pc` target to fail - it does not even use this code

**Comment by damianloew** [comment]
> I don't think my change is causing `host-generic-pc` target to fail - it does not even use this code

There was a failed build on nucleo (before splitting test-libc) caused by some GH problems (The hosted runner: GitHub Actions 9 lost communication with the server) and after some time you probably clicked `re-run failed jobs`, but during this period splitting test-libc binary has been merged and the previous builds were used (without test-libc-pthread binary). Clicking `re-run all jobs` triggered by @mateusz-bloch finally helped. So it seems to be fine as we can see in `checks` tab now.

**Review by badochov — DISMISSED** [review]

File: stdio/file.c [hunk 3]
```diff
@@ -857,45 +857,53 @@ int ungetc(int c, FILE *stream)
...
 {
-	int len = strlen(s), l = 0, err;
+	size_t len = strlen(s);
+	size_t wrote = 0, ret;
 
-	while (l < len) {
-		if ((err = fwrite_unlocked((void *)s + l, 1, len - l, stdout)) < 0) {
-			return -1;
+	do {
+		ret = fwrite_unlocked(s + wrote, 1, len - wrote, stream);
```
[the remark this answers is not in the document]
Comment (agkaminski) [reply stdio/file.c:866]: Right, now it should be ok, thanks

File: stdio/file.c [hunk 4]
```diff
@@ -857,45 +857,53 @@ int ungetc(int c, FILE *stream)
...
+int fputs_unlocked(const char *s, FILE *stream)
 {
-	int len = strlen(s), l = 0, err;
+	size_t len = strlen(s);
+	size_t wrote = 0, ret;
 
-	while (l < len) {
-		if ((err = fwrite_unlocked((void *)s + l, 1, len - l, stdout)) < 0) {
-			return -1;
+	while (wrote < len) {
```
Comment (badochov) [inline stdio/file.c:865]: What do you think about using `for` here?

**Review by badochov — APPROVED** [review]

**Pull request merged**